### ISSUE TITLE
Orchestration - Feature: Support Stack lookup by Stack ID (only) or Stack Name (only)

### ISSUE THREAD

**Issue by scott-chaney** [gophercloud/gophercloud#1157]
Currently the Orchestration (Stack) Gophercloud implementation only supports stack Get() lookup by Name + ID.  Additionally, the List() lookup does not support the ID ListOpts parameter, which is available on the Openstack API.  It would be useful to be able to lookup all stack details when starting with just a stack name or a stack id.

The Stack ID is the only stack parameter that has a uniqueness constraint in the DB, but the API does enforce uniqueness per project of the stack name as well.

After doing some testing with Gophercloud and the Openstack API, I find that we could just update the current stack Get() function to take an "ident" which can be either the stack name or stack id, the API appears to work fine with either (since the project ID is required).  It's probably not a good idea to change a public method, so maybe we can just allow an empty string for either one of the parameters?

If it is preferred to not make any changes to how those public methods work, it might be fine just to support the "id" field in the list opts.  I did a quick test on this and it appears to work well, though I imagine that this is less efficient on the Openstack side.

I am happy to work on this issue if it gets the thumbs up.

Related Openstack documentation:

- [Get stack by ID](https://github.com/gophercloud/gophercloud/blob/master/openstack/orchestration/v1/stacks/requests.go#L220)

- [List stacks](https://developer.openstack.org/api-ref/orchestration/v1/index.html#list-stacks) <- note the "id" request parameter

- [show() definition](https://github.com/openstack/heat/blob/d899c240e49a72246926b46f40f8076159bb5196/heat/api/openstack/v1/stacks.py#L429)

Related gohpercloud code:

- [ListOpts parameters (missing 'id')](https://github.com/gophercloud/gophercloud/blob/master/openstack/orchestration/v1/stacks/requests.go#L220)

- [Stack Get() function](https://github.com/gophercloud/gophercloud/blob/master/openstack/orchestration/v1/stacks/requests.go#L257)

**Comment by jtopjian** [comment]
With regard to "Get":

I think Gophercloud's current `Get()` function might be implementing this action:

https://developer.openstack.org/api-ref/orchestration/v1/index.html#show-stack-details

The API documentation for the `/v1/{tenant_id}/stacks/{stack_identity}` URL is titled as "Find stack", so maybe we can add a new function called `Find(client, ident)` ?

We'll need to see a link to the Heat API code which implements the action.

With regard to "List":

If the API accepts an ID query arg, then we can absolutely add it. We'll need to see the Heat API code which confirms, this, though. 

Additionally, the API docs specify that the fields can be specified multiple times. If this is true (again, referenced in the API code), we should change the `ListOpts` fields to `[]string` where applicable. This will cause backwards-incompatible changes, but that's OK. In general, breaking changes are fine in Gophercloud. I'm definitely OK with them for Orchestration since we haven't paid much attention to those packages and it can use a good number of updates/fixes.

Let me know if this makes sense and/or if you have any questions.

**Comment by scott-chaney** [comment]
Thanks for the followup @jtopjian 

Adding the Find() function sounds like a good idea.

The documentation and the code appear to have some inconsistencies.  The term "stack_identity" in the API doc can apparently refer to either the stack_name or the stack_id (both are unique per project) but the code refers to just "stack_name" in this case.  The links to the code below show the "stack_name" in the "find" call being resolved to the name of the stack or the UUID of the stack.

**Openstack code for stack Find():**
See the [lookup](https://github.com/openstack/heat/blob/master/heat/api/openstack/v1/__init__.py#L156) route definition
and the associated [controller function](https://github.com/openstack/heat/blob/master/heat/api/openstack/v1/stacks.py#L421)

[RPC ](https://github.com/openstack/heat/blob/master/heat/rpc/client.py#L128)
Which calls to this [Service Function](https://github.com/openstack/heat/blob/07fa426a16c11e1de7a014eb8f9ca016dbe38d8f/heat/engine/service.py#L435) which processes the input for UUID or Name

The final result is the redirect hint which the Gophercloud providerClient's http.Client should handle fine.

**Openstack code for stack List():**
[Routing](https://github.com/openstack/heat/blob/master/heat/api/openstack/v1/__init__.py#L129)
[controller function](https://github.com/openstack/heat/blob/master/heat/api/openstack/v1/stacks.py#L207)
The filters are then passed all the way though to the DB query.
[RPC](https://github.com/openstack/heat/blob/master/heat/rpc/client.py#L138)
[Service](https://github.com/openstack/heat/blob/07fa426a16c11e1de7a014eb8f9ca016dbe38d8f/heat/engine/service.py#L523)
etc.. until The [DB layer](https://github.com/openstack/heat/blob/37336c0faf5df57ea878e104ae6cc7fcd01509c1/heat/db/sqlalchemy/api.py#L709)

I did a quick check by making the related changes on the Gophercloud side, and it seems to work well.

I'll get working on a real implementation, test it out, and send an update.  Thanks!

**Comment by ricolin** [comment]
IMO, what we should do is allowing Get, and Delete to provide either name or id. Since Name and ID are both identical(that's why our API allow provide either name or id)

As for Update, and any actions can provide either name or id, and add a Find function to find Name from List (it's idential, so you not supposed to find any duplicated stack name) and translate it to ID. At least that's how we deal this in Python SDK

**Comment by jtopjian** [comment]
@ricolin Right, however, we have already defined `Get` as the [Show stack details](https://developer.openstack.org/api-ref/orchestration/v1/index.html#show-stack-details) API call:

```
/v1/{tenant_id}/stacks/{stack_name}/{stack_id}
```

So in order for us to define `Get` as [Find stack](https://developer.openstack.org/api-ref/orchestration/v1/index.html#find-stack)

```
/v1/{tenant_id}/stacks/{stack_identity}
```

We will need to rename the existing `Get` function as something like `GetDetails`. That's a significant breaking change, but I would be OK with it if it's the best solution.

**Comment by scott-chaney** [comment]
@ricolin, @jtopjian Thanks for thinking on this.  I am not tied to either option but am hesitant to make a breaking change. However, just needing a single parameter for any of the mentioned requests would definitely make it more simple to use.

@jtopjian if I am understanding the suggestion from @ricolin, we would not really need to implement the `GetDetails` method, since the `Get `would just take either parameter and provide the same functionality.  Or maybe you were getting at keeping `GetDetails` would be good if the overhead from the redirect from the Find route is of any concern?

If we fully adopt the pattern that the python heatclient employs, I believe it would mean breaking changes for all methods that currently require both parameters in gophercloud. If we are cool with that then I will move forward on that path.

[Python heatclient implementation](https://github.com/openstack/python-heatclient/blob/master/heatclient/v1/stacks.py#L270) for reference.

I haven't had a chance to circle back on this yet, hopefully in a week or so.

**Comment by jtopjian** [comment]
> we would not really need to implement the `GetDetails` method, since the `Get` would just take either parameter and provide the same functionality. Or maybe you were getting at keeping `GetDetails` would be good if the overhead from the redirect from the Find route is of any concern?

Good point and good question. I don't have enough experience with the Heat API to make a confident call on that, so I'll defer to you folks :)

My understanding right now is that the Heat API has two `GET` routes/calls:

```
/v1/{tenant_id}/stacks/{stack_identity}
/v1/{tenant_id}/stacks/{stack_name}/{stack_id}
```

Gophercloud implements the second, but we want to implement the first. The question boils down to: do we drop support for the current route altogether, rename it, or implement the first style as a different name other than `Get`?

If I understand correctly, both `stack_name` and `stack_id` are unique identifiers that will point to the same stack.

So one thing I want to understand is _why_ the second form exists?

If it's a legacy thing, then I would be in favor of renaming it to something else, mostly for those devs who need to use that style.

As I've mentioned before, Gophercloud's orchestration packages need a good amount of updating, so right now I'm OK with significant / backwards-incompatible changes _if_ those changes are the best way to get a good Orchestration implementation in Gophercloud.

> I haven't had a chance to circle back on this yet, hopefully in a week or so.

No problem at all! Take your time.

**Comment by ricolin** [comment]
> /v1/{tenant_id}/stacks/{stack_identity}
This API is use to find a stack [1], so IMO we can use `Find` as name of function
> /v1/{tenant_id}/stacks/{stack_name}/{stack_id}
This API (which we implemented with func `Get`) will get details.

> 
> So one thing I want to understand is _why_ the second form exists?
Because when people start planning on the first day, they don't think to put parameters into URI is a bad idea:)

Another action that we should do here is to try to reduce any chances that asking a user to provide both ID and Name of a Stack. Which we should be able to do a stack list before get/find and translate the providing parameter into name and ID (like what we do in Python SDK).


I will try to look into this together:)

[1] https://developer.openstack.org/api-ref/orchestration/v1/#find-stack

**Comment by scott-chaney** [comment]
@jtopjian and @ricolin I submitted [This PR](https://github.com/gophercloud/gophercloud/pull/1287) for the simple Find() case without touching the other public methods.

> Another action that we should do here is to try to reduce any chances that asking a user to provide both ID and Name of a Stack. Which we should be able to do a stack list before get/find and translate the providing parameter into name and ID (like what we do in Python SDK).

Since the "Find" operation returns a redirect that is a direct lookup of the stack, maybe we should keep the existing "Get" method if the user has the name and id and does not want to incur the extra lookup.

Let me know if there are any improvements or changes that you guys want to explore before we consider merging.

**Comment by hardys** [comment]
> So one thing I want to understand is _why_ the second form exists?

I was involved with Heat development back when this was added (about 6 years ago so it's hard to remember exactly!) but IIRC it was a combination of uncertainty over the token scope e.g could we always rely on project scoped tokens and thus name uniqueness, the ability to retrieve deleted stacks (where the name is no longer necessarily unique), and also general inconsistency between the various other OpenStack APIs at that time.

With the benefit of hindsight it's a historical mistake, and probably we should rev the API to align with most other projects, e.g just have the stack ID in the URI path and rely on a list query with the name to get the ID.  However the redirect has been working well enough for $years and we've not had enough other reasons to put users through the pain of an API bump.

@zaneb may have other recollections but for now I'd agree with the previous discussion, e.g use of the Find/Redirect to enable lookup by name/id is probably the best plan.

**Comment by zaneb** [comment]
tl;dr it seemed like a good idea at the time.

There wasn't a consistent standard within OpenStack at the time (2012). The idea was that within a tenant we wanted stacks to be uniquely identifiable by name, and I figured most people would access them that way. I also wanted the URLs to be human-readable. So we put the tenant id followed by the name in the URL. However, once a stack is deleted it is still accessible in the DB, so I put the UUID after that so that you could still access an old one if you knew the UUID. If you only knew the name, the idea was that you would start with:

    /v1/{tenant_id}/stacks/{stack_name}

and it would redirect you to the canonical URL:

    /v1/{tenant_id}/stacks/{stack_name}/{stack_id}

Immediately upon writing the CLI client we realised that this presented a problem: you generally got either the stack id or the stack name as input, but never both. So we added the ability to also get a redirect from:

     /v1/{tenant_id}/stacks/{stack_id}

The rest of the story is as Steve said. There's now a kind-of standard around not putting the tenant_id (now called project_id) in the URL, and only putting the UUID in the URL. In light of the the current design is clearly a mistake, but the pain for users of revving the API has always outweighed the pain of continuing to do it this way.

**Comment by jtopjian** [comment]
@hardys @zaneb Thank you both very much for the detailed history!

It sounds like the best way to go is to retain the current implementation and add a new function (`Find`?) which implements the single unique identifier argument.

**Comment by jtopjian** [comment]
Looks like this was implemented in #1287
